casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -0,0 +1,280 @@
+    /**
+      * Blocks received and stored in BlockStore with parent relations
+      */
+    blocksSt: Map[MId, Set[MId]],
+    /**
+      * Blocks receiving status
+      */
+    receiveSt: Map[MId, RecvStatus],
+    /**
+      * Blocks mapping with children relations
+      */
+    childRelations: Map[MId, Set[MId]]
+) {
+
+  /**
+    * Keep status of block as received process started
+    *  - thread safe sync like opening transaction for a block by block hash
+    */
+  def beginReceived(id: MId): (BlockReceiverState[MId], Boolean) = {
+    val expectedReceive = receiveSt.get(id).collect { case PendingRequest => true }.getOrElse(true)
+    // If state is not known or pending it's request, continue with receiving
+    if (expectedReceive) {
+      // Update state to begin received status
+      val newReceiveSt = receiveSt + ((id, BeginReceived))
+      (copy(receiveSt = newReceiveSt), true)
+    } else {
+      (this, false)
+    }
+  }
+
+  /**
+    * End of block receiving status update
+    *  - like closing transaction, block received and stored
+    *
+    *  @return unseen parent dependencies
+    */
+  def endReceived(id: MId, parents: List[(MId, Boolean)]): (BlockReceiverState[MId], Set[MId]) = {
+    val curStateOpt = receiveSt.get(id)
+    assert(
+      curStateOpt == BeginReceived.some,
+      s"Received should be called only in begin received state, actual: $curStateOpt, hash: $id"
+    )
+    curStateOpt
+      .collect {
+        case BeginReceived =>
+          // Update blocks state, keep unseen parents only
+          val parentsNotStored = parents.filter(_._2).map(_._1).toSet
+          val unseenParents    = parentsNotStored -- blocksSt.keySet -- receiveSt.keySet - id
+          val newBlocksSt      = blocksSt + ((id, unseenParents))
+
+          // Update block status to received and set unseen parents to Pending receive state
+          val newReceiveStored  = receiveSt + ((id, ReceivedAndStored))
+          val newPendingReceive = unseenParents.map((_, PendingRequest))
+          val newReceiveSt      = newReceiveStored ++ newPendingReceive
+
+          // Update children relations of received block
+          val parentsHashes = parents.map(_._1)
+          val newChildRelations = parentsHashes.foldLeft(childRelations) {
+            case (acc, parent) =>
+              val childs = acc.getOrElse(parent, Set())
+              acc + ((parent, childs + id))
+          }
+
+          // New state
+          val newState = copy(
+            blocksSt = newBlocksSt,
+            receiveSt = newReceiveSt,
+            childRelations = newChildRelations
+          )
+
+          (newState, unseenParents)
+      }
+      .getOrElse((this, Set()))
+  }
+
+  /**
+    * Finished block validation, update state and return next blocks
+    *
+    * @return next blocks with validated dependencies
+    */
+  def finished(id: MId, parents: Set[MId]): (BlockReceiverState[MId], Set[MId]) = {
+    val parentsInState = blocksSt.get(id)
```

Comment:
It looks to me that you used 3 different approaches in `beginReceived`, `endReceived` and `finished`
if-else, option with assertion and products of 2 options, correspondingly.

I think it all can be unified as this
```scala
val validTransition = 
   blocksSt.get(id).isEmpty && receiveSt.get(id).collect { case PendingRequest => true }.getOrElse(true) // beginReceived
   blocksSt.get(id).isEmpty && receiveSt.get(id).contains(BeginReceived) // endReceived
   blocksSt.get(id).exists(_.isEmpty) && receiveSt.get(id).contains(PendingValidation) // finished
if (validTransition) {
   // do
}
else ((this, Set())) // or throw exception
```

Reply:
> Why case ReceivedAndStored => is here? finished should be called on block which is PendingValidation only?

Block can be downloaded (in ReceivedAndStored state) and have all justifications already in the DAG. This situation is not tracked by the receiver. This simplify the solution because there is no need to update state in this case.

This can be optimized and manage this inside the state, it will reduce the number of DAG contains checks for block justifications.